### PR TITLE
JAMES-3369 Enhance EmailGetSerializer

### PR DESCRIPTION
**Pull request by chibenwa**
- Avoid nested serializer definition - this enables a single evaluation

### CODE REVIEW

**Comment by chibenwa** [comment]
## Why

We notice `Email/get` involving body fields is significantly slower on RFC-8621 compared to draft (factor 2)

We suspected the recursivity in the JSON serializer to cause issues here.

## Before

Capture d’écran de 2020-11-06 14-10-07

## After

Capture d’écran de 2020-11-06 14-12-49

## Draft

Capture d’écran de 2020-11-06 14-13-21

## Conclusion

While mean time of "openMessage" is still 2x time lower on new implem compared to draft, we are now on-par on p99 on both implems.

This changeset unlocks easy persormance enhencements upon very low load for the new RFC implem.

We will likely benefit from a full review of our RFC-8621 JSON stack.

**Review by rouazana — COMMENTED** [review]
I"m not enough fluent in Scala to see the difference, can you explain me please?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala [hunk 1]
```diff
@@ -89,50 +89,49 @@ object EmailGetSerializer {
...
+    JsPath.write[Map[String, Option[EmailHeaderValue]]]) (unlift(EmailFastView.unapply))
+  private val emailHeaderViewWrites: OWrites[EmailHeaderView] = (JsPath.write[EmailMetadata] and
+    JsPath.write[EmailHeaders] and
+    JsPath.write[Map[String, Option[EmailHeaderValue]]]) (unlift(EmailHeaderView.unapply))
+  private val emailMetadataViewWrites: OWrites[EmailMetadataView] = view => Json.toJsObject(view.metadata)
 
   private def emailWritesWithPropertyFilter(properties: Properties)(implicit partsWrites: Writes[EmailBodyPart]): Writes[EmailView] = {
-    implicit val emailMetadataWrites: OWrites[EmailMetadata] = Json.writes[EmailMetadata]
-    implicit val emailHeadersWrites: Writes[EmailHeaders] = Json.writes[EmailHeaders]
     implicit val emailBodyWrites: Writes[EmailBody] = Json.writes[EmailBody]
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala:127]: Could you move this line out of the method in order not to do it at each serialization ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala [hunk 2]
```diff
@@ -89,50 +89,49 @@ object EmailGetSerializer {
...
-    implicit val emailBodyMetadataWrites: Writes[EmailBodyMetadata] = Json.writes[EmailBodyMetadata]
 
     val emailFullViewWrites: OWrites[EmailFullView] = (JsPath.write[EmailMetadata] and
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala:129]: idem

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala [hunk 3]
```diff
@@ -89,50 +89,49 @@ object EmailGetSerializer {
...
-    val emailFastViewWrites: OWrites[EmailFastView] = (JsPath.write[EmailMetadata] and
-        JsPath.write[EmailHeaders] and
-        JsPath.write[EmailBodyMetadata] and
-        JsPath.write[Map[String, Option[EmailHeaderValue]]]) (unlift(EmailFastView.unapply))
-    val emailHeaderViewWrites: OWrites[EmailHeaderView] = (JsPath.write[EmailMetadata] and
-        JsPath.write[EmailHeaders] and
-        JsPath.write[Map[String, Option[EmailHeaderValue]]]) (unlift(EmailHeaderView.unapply))
-    val emailMetadataViewWrites: OWrites[EmailMetadataView] = view => Json.toJsObject(view.metadata)
-
     val emailWrites: OWrites[EmailView] = {
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala:135]: idem

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala:127]: No it depends on the implicit partsWrites.... (which may recurse, and be filtered itself...)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala:129]: idem

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailGetSerializer.scala:135]: idem

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
> I"m not enough fluent in Scala to see the difference, can you explain me please?

Well you have format definition, wich can be doing statically upon program init, and the the actual serialization hapening at runtime.

We were (and still are, but less) doing some format definition at runtime, for each request.

The thing is, some wizardry, at some points will be needed, as the serialisation logic is dynamic in jmap. Think property filtering, extension filtering.

Maybe profiling can help better identify miss-usage than this blind educated guess pr...

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Demonstrating the impact of `def` usage within JSON context

## Before https://github.com/linagora/james-project/pull/4008/commits/f9824051ac4164daa1e8a70d97d0d26d7573be12

Capture d’écran de 2020-11-09 14-24-43

## After https://github.com/linagora/james-project/pull/4008/commits/f9824051ac4164daa1e8a70d97d0d26d7573be12

Capture d’écran de 2020-11-09 14-25-24

**Comment by chibenwa** [comment]
Edit: we will need to adapt all property filtering to use **transformers** instead of implicit changes.

**Review by remk — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
The above performance tests gains proved to be premature.

We noticed bad performance arising repeatedly independently of james-code / bench code, leading to (really) downgraded / failing perfs.

Repeated runs leads to less advantageous (yet beneficial) improvements

## draft

draft

## RFC-8621 before JSON rework

before

## RFC-8621 after JSON rework

after

## Conclusion

 - JSON enhancements leads to significant improvements compared to RFC-8621 base implementation, and needs to be carried over.

 - RFC-8621 suffers from significantly worst mean times than Draft

 - However it benefits from lower p99 & standard deviation, thus having a more stable behaviour.

Other performance measurements will be needed once this JSON rework is conducted out (and cleaned up!) both with 600 user/h and 3600 user/h.

As such, performance of RFC-8621 do not seem to be a hard blocker for adoption.

**Review by Arsnael — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Does this still need more work? Or can we potentially merge it? It seems enhancing perfs already, maybe further changes could be made in other PRs?

**Comment by chibenwa** [comment]
> Or can we potentially merge it?

Yes.

**Comment by Arsnael** [comment]
> > Or can we potentially merge it?
> 
> Yes.

Undraft it and rebase it on master then please :)

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**